website/blog/2022-11-31-simple-2.mdx

```diff
@@ -0,0 +1,115 @@
+
+Such an approach used to make simple functions significantly slower, up to more than `10X` for input maps. In order to mitigate those regressions, Velox introduced <a href="https://github.com/facebookincubator/velox/blob/main/velox/expression/ComplexViewTypes.h">View types</a>: ArraViews, MapViews ...etc. The goal is to keep the authoring simple but achieve **at least** the performance of a basic vector implementation that decodes input and applies some logic for every row without any special optimizations.
+
+The new view types are **Lazy**, very cheap to construct and do not materialize the underlying data unless the code accesses it. They are also designed to have interfaces similar to those of std::containers.
+
+In a simplified form, an ArrayView stores the length and the offset of the array within the vector, in addition to a pointer to the elements array. Only when an element is accessed then an OptionalAccessor is created, which contains the index of the accessed element and a pointer
+to the containing vector. Only when the user calls value() or has_value() on that accessor then the nullity or the value is read. Other view types are implemented in a similar way,
+The graph below illustrates the process.
+
+<figure>
+ <img src="/img/simple1_4.png"/>
+</figure>
+
+
+The graph below compares the runtime of some functions written in the simple interface before and after the introduction of the view types. The speedup for arrays is around 2X, for maps the speed is much higher > `10X` because materializing the intermediate representation previously involves hashing the elements while constructing the hashmap. Furthermore, the overhead of materialization for nested complex types is very high as well, as reflected in _row_arrays_sum_.
+
+<figure>
+ <img src="/img/simple1_6.png"/>
+ <figcaption> Runtimes of functions before and after the introduction of view types, normalized to the runtime of the version that uses the view types.</figcaption>
+</figure>
+
+
+The graph below compares the runtimes of some functions written using the simple interface, a basic vector function implementation with no special optimizations for the non general case, and a vector implementation that is specialized for flat and null free. The bars are annotated with the line of codes (LOC) used to implement each function.
+
+
+<figure>
+ <img src="/img/simple1_7.png"/>
+</figure>
+
+We can see that LOC are significantly lower for simple functions. ArraySum with flat and null free optimization is faster because the summation can be optimized much better when it's performed over a sequential array of data.
+The reason the simple is faster than the vector for some benchmarks is because we have several optimizations in the simple interface such as vector-reuse, bulk null setting ..etc,  that are not implemented in the basic vector versions. We will talk about those optimizations in a future post.
```

Comment:
Maybe we can leave a link to the VLDB paper, where we also discuss and present some of these results?

Reply:
oh we alreasy talked about those in post 1 in detauls let me link it